### DISCUSSION TITLE
Back to newbie state... controlled vs uncontrolled

### DISCUSSION THREAD

**Discussion by msageryd** [react-hook-form/react-hook-form#7566] [Q&A]
My forms works great. They are performant and they do exactly as I want. But I fear that I have messed up and/or misunderstood some core concepts.

Controlled vs uncontrolled, where do you draw the line?

This is what I do (React-Native):

1. register all my fields with `register` (I need all fields registered even if not every field is visible for now, due to some cross field rules I need to process)
2. my custom input components uses `useController` to get hold of `value` and `onChange` which I inject in the actual input component (I'm using components from react-native-paper). 

The above seems to work perfectly. I also was pleasantly surprised to see that I could override register-stuff with useController-stuff. I.e. if I register a field with "isRequired = true", I can override this when I mount a field with useController.

**But..** I suspect that I now have a mix of controlled (the mounted components) and uncontrolled fields. Is this a bad thing? Should I do it in any other way?
Should I for example keep everything uncontrolled? In that case I suppose I should use setValue instead of onChange?

**Odd, but good behavior**
In this discussion: https://github.com/react-hook-form/react-hook-form/discussions/7557 I realized that my forms behave differently than the form @moshyfawn put together in CSB. 

- If I mount two different input fields with the same field name, they update eachother. It's kind of odd to mount the same field more than once, but I actually have a huse case for ot, and it works perfectly.

Since Moshyfawn's form example does not behave like this, I'm now worried that I have done something wrong.

N.B. apart from my forms bahavior (i.e. exactly as I want), they are also performant, with minimal rerenders.

**Comment by msageryd** [comment]
Thank you @Moshyfawn for clarifying a lot. I'm trying to answer this question myself and I hope someone could chime in and correct me if I'm out of whack.

Maybe my explanation below is correct, or could be correct after some colaboration. In that case I think it would fit well in the FAQ. I can imagine more than me having a hard time to grasp controlled vs uncontrolled.

My take (guess)

**Controlled:**
`onChange` connects the input with RHF and lets RHF get hold of all data entered in the input. RHF stores the value, possibly in a converted form (i.e. numeric).

`value` is also fed back to the component from RHF, possibly after some validation/formatting/conversion. If values are not stored as strings in RHF they will be fed back to the component converted to strings, since the components local state is string.


**Uncontrolled:**
`onChange` connects RHF with the component in the same manner as for a controlled component.

No value is fed back to the component, hence there is no opportunity to format the value.

The component will store the string-version of all input in local state and RHF will store the possibly converted value. We will have "two sources of truth".

What if an uncontrolled field gets altered in RHF via setValue from an externa event? 
The input would not show the new value, would it? If this is correct, I feel that uncontrolled fields should be used cautiously.

**Reply by bluebill1049** [reply]
> No value is fed back to the component, hence there is no opportunity to format the value.

look into `setValue` option.

**Reply by bluebill1049** [reply]
> What if an uncontrolled field gets altered in RHF via setValue from an externa event?

react hook form have an internal logic to info you of the state update and reconcile for state update if required.
